src/rtStorage.cpp

```diff
@@ -0,0 +1,461 @@
+      }
+
+      a->pushBack(row);
+      rc = sqlite3_step(stmt);
+
+      rtLogInfo("%s: key '%s'", __FUNCTION__, row.get<rtString>("key").cString());
+    }
+
+    retValue = a;
+
+    sqlite3_finalize(stmt);
+  }
+
+  return RT_OK;
+}
+
+rtError rtStorage::removeItem(const char* key)
+{
+  if (!key || *key == 0)
+    return RT_ERROR_INVALID_ARG;
+
+  rtLogInfo("%s: '%s'", __FUNCTION__, key);
+
+  sqlite3* &db = SQLITE;
+
+  if (db)
+  {
+    sqlite3_stmt *stmt;
+    sqlite3_prepare_v2(db, "DELETE FROM ItemTable where key = ?;", -1, &stmt, NULL);
+
+    sqlite3_bind_text(stmt, 1, key, -1, SQLITE_TRANSIENT);
+
+    int rc = sqlite3_step(stmt);
+    if (rc != SQLITE_DONE) 
+    {
+      rtLogError("ERROR removing data: %s", sqlite3_errmsg(db));
+    }
+
+    sqlite3_finalize(stmt);
+  }
+
+  updateSize();
+
+  return RT_OK;
+}
+
+// Methods for managing current size and quota
+rtError rtStorage::calculateCurrentSize(uint32_t &size) const
+{
+  sqlite3* &db = SQLITE;
+
+  size = 0;
+
+  if (db)
+  {
+    sqlite3_stmt *stmt;
+    sqlite3_prepare_v2(db, "SELECT sum(length(key)+length(value)) FROM ItemTable;", -1, &stmt, NULL);
+
+    int rc = sqlite3_step(stmt);
+    if (rc == SQLITE_ROW)
+    {
+      size = sqlite3_column_int(stmt, 0);
+    }
+    sqlite3_finalize(stmt);
+  }
+
+  return RT_OK;
+}
+
+rtError rtStorage::setCurrentSize(const uint32_t size)
+{
+  sqlite3* &db = SQLITE;
+
+  if (db)
+  {
+    sqlite3_stmt *stmt;
+    sqlite3_prepare_v2(db, "UPDATE SizeTable SET currentSize = ?;", -1, &stmt, NULL);
+
+    sqlite3_bind_int(stmt, 1, size);
+
+    int rc = sqlite3_step(stmt);
+    if (rc != SQLITE_DONE) 
+    {
+      rtLogError("ERROR inserting size: %s", sqlite3_errmsg(db));
+    }
+
+    sqlite3_finalize(stmt);
+  }
+
+  return RT_OK;
+}
+
+rtError rtStorage::getCurrentSize(uint32_t &size) const
+{
+  sqlite3* &db = SQLITE;
+
+  size = 0;
+
+  if (db)
+  {
+    sqlite3_stmt *stmt;
+    sqlite3_prepare_v2(db, "SELECT * FROM SizeTable;", -1, &stmt, NULL);
+
+    int rc = sqlite3_step(stmt);
+    if (rc == SQLITE_ROW)
+    {
+      size = sqlite3_column_int(stmt, 0);
+    }
+    sqlite3_finalize(stmt);
+  }
+
+  return RT_OK;
+}
+
+rtError rtStorage::updateSize()
+{
+  uint32_t size = 0;
+  rtError retStat = calculateCurrentSize(size);
+
+  if (RT_OK == retStat)
+  {
+    retStat = setCurrentSize(size);
+    if (RT_OK == retStat)
+    {
+      rtLogDebug("storage size is %d", size);
+    }
+    else
+    {
+      rtLogError("failed to update size");
+    }
+  }
+  else
+  {
+    rtLogError("failed to calculate size");
+  }
+
+  return retStat;
+}
+
+rtError rtStorage::verifyQuota(const char* key, const rtValue& value) const
+{
+  if (!key || *key == 0)
+    return RT_ERROR_INVALID_ARG;
+
+  uint32_t size = 0;
+  rtError retStat = getCurrentSize(size);
+
+  if (RT_OK == retStat)
+  {
+    rtString keyStr = key;
+    rtString valueStr = value.toString();
+    uint32_t newSize = size + keyStr.byteLength() + valueStr.byteLength();
```

Comment:
This does not seem to account for the size difference if key already exists with a different value.